### ISSUE TITLE
Using .default with z.object doesn't do anything, and neither does .required  if z.object has defaults

### ISSUE THREAD

**Issue by gotofritz** [colinhacks/zod#3482]
Hi, thanks for the great library.

I am to have two version of a schema, one with the fields optional and default values, the other with all fields required.

If I start from the optional version...

```js
var allOptionals = z.object({
  hero: z.string().default("rick"),
  sidekick: z.string().default("morty"),
})
var allRequired = allOptionals.required();

allRequired.parse({});
// expected: error
// actual: { hero: "rick", sidekick: "morty" }
```

and if I start from the required version

```js
var allRequired = z.object({
  hero: z.string(),
  sidekick: z.string(),
})
var allOptionals = allRequired.default({
  hero: "rick",
  sidekick: "morty",
});

allOptionals.parse({});
// expected: { hero: "rick", sidekick: "morty" }
// actual: error
```

How can I add a bunch of defaults to a `z.object` in one go, without definiing them individually? (I get them from an api call)

EDIT: I am aware of .transform and .refine, but they return a ZodEffect and not a ZodObject. I also need to use those schemas in .merge. I could use sourceType or innerType but it still doesn't come our right. 

Thanks

**Comment by galaxyblur** [comment]
I agree, this doesn't work as expected, and I'll add that even `myObjectSchema.parse(undefined)` does not return the default values. If `default` is not useable on an object schema, it should be unavailable there and documented as such.
